### DISCUSSION TITLE
Difference b.w calling `self.log(..., on_step=False, on_epoch=True)` in training_step() and training_epoch_end()

### DISCUSSION THREAD

**Discussion by janghyuk-choi** [ashleve/lightning-hydra-template#485] [Q&A] [answered]
https://github.com/ashleve/lightning-hydra-template/blob/75b44fff041dc15e7b4e5d54c16aa906531c86a2/src/models/mnist_module.py#L75-L76

In the src/models/mnist_module.py, `self.log()` is called inside `training_step()` with `on_step=False` and `on_epoch=True`. 

I wonder what the difference is compared to using `self.log(..., on_step=False, on_epoch=True)` inside `training_epoch_end()`.

It seems that the resulting values in either way are the same.

Is there any difference in the internal mechanism?

**Comment by ashleve** [comment, marked as answer]
No difference as far as I'm aware, as long as we're logging torchmetrics object directly. Not sure if there won't be slight differences when we log through value though.

- Inside `training_epoch_end()` flags `on_step=False, on_epoch=True` don't matter as you're always logging only once per epoch.
- Inside `training_step()` setting `on_epoch=True` will make lightning average given value over logs from all steps.

**Reply by janghyuk-choi** [reply]
Thank you for your kind reply!

It seems like just a convention by [PL](https://www.pytorchlightning.ai).

And there has been a discussion on the purpose or function of that code (https://github.com/Lightning-AI/lightning/issues/3228).

**Reply by adosar** [reply]
> No difference as far as I'm aware, as long as we're logging torchmetrics object directly. Not sure if there won't be slight differences when we log through value though.
> 
> * Inside `training_epoch_end()` flags `on_step=False, on_epoch=True` don't matter as you're always logging only once per epoch.
> * Inside `training_step()` setting `on_epoch=True` will make lightning average given value over logs from all steps.

Is this behavior somewhere documented that "on_epoch=True" doesn't matter for "on_train_epoch_end"?
